#### CV of Hash Features

In [1]:
import sys
sys.path.insert(0, '../../bosch_helper')
from bosch_helper import *

%matplotlib inline

/c3se/NOBACKUP/users/lyaa/conda_dir/miniconda/envs/kaggle/lib/python3.7/site-packages/sklearn/utils/__init__.py:4: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
  from collections import Sequence


##### Load data up to benchmark 7

In [2]:
filename = pd.HDFStore('../hpopt_1/numeric_b1_b7_nf149.hdf')
print(filename.keys())
filename.close()

['/numeric', '/y_train']


In [3]:
x = pd.read_hdf('../hpopt_1/numeric_b1_b7_nf149.hdf', 'numeric')
y_train = pd.read_hdf('../hpopt_1/numeric_b1_b7_nf149.hdf', 'y_train')

##### Load hash feature

In [8]:
nhash = pd.read_hdf('hash_numeric.hdf', 'hash_numeric')

In [10]:
x = x.join(nhash)

In [12]:
x_train = x.loc['train']
x_test = x.loc['test']

In [17]:
del x, nhash
gc.collect()

28

### CV score based on stratified KFold with repeated models

In [ ]:
# CV
# specify parameters 
# 'booster':'gbtree'
params = {'max_depth':14, 'eta':0.03, 'silent':1, 'objective':'binary:logistic', 'nthread':20,
         'lambda':4, 'subsample':0.9, 'min_child_weight':5, 'booster':'gbtree', 'alpha':0,
         'base_score':0.0058, 'colsample_bytree':0.6}

# 'booster':'dart'
# params = {'max_depth':14, 'eta':0.03, 'silent':1, 'objective':'binary:logistic', 'nthread':20,
#         'lambda':4, 'subsample':0.9, 'min_child_weight':5, 'booster':'dart', 'alpha':0,
#         'base_score':0.0058, 'nthread':20, 'colsample_bytree':0.6, 
#         'sample_type':'uniform', 'normalize_type':'tree', 'rate_drop':0.1, 'skip_drop':0.2, 'one_drop':True}

cv_results, clfs, running_time = cross_val_predict_skf_rm_xgb(params, x_train, y_train, 
                                                              num_boost_round=80, n_splits=5, 
                                                              n_repeats=3, random_state=5870577, 
                                                              verbose_eval=True)

results = {'clfs': clfs, 'cv_results': cv_results, 'running_time': running_time}
save_pickle(results, 'results_benchmark_8_cv_5_hash.pickle')

[0]	train-error:0.005811	test-error:0.005812	train-MCC:0.309215	test-MCC:0.301967
[1]	train-error:0.005811	test-error:0.005812	train-MCC:0.335733	test-MCC:0.326261
[2]	train-error:0.005811	test-error:0.005812	train-MCC:0.348581	test-MCC:0.33397
[3]	train-error:0.005811	test-error:0.005812	train-MCC:0.361706	test-MCC:0.351471
[4]	train-error:0.005811	test-error:0.005812	train-MCC:0.386493	test-MCC:0.365614
[5]	train-error:0.005811	test-error:0.005812	train-MCC:0.393544	test-MCC:0.37018
[6]	train-error:0.005811	test-error:0.005812	train-MCC:0.409499	test-MCC:0.382208
[7]	train-error:0.005811	test-error:0.005812	train-MCC:0.41754	test-MCC:0.390952
[8]	train-error:0.005811	test-error:0.005812	train-MCC:0.427532	test-MCC:0.399954
[9]	train-error:0.005811	test-error:0.005812	train-MCC:0.431908	test-MCC:0.399703
[10]	train-error:0.005811	test-error:0.005812	train-MCC:0.437134	test-MCC:0.405093
[11]	train-error:0.005811	test-error:0.005812	train-MCC:0.445974	test-MCC:0.406293
[12]	train-error:

[19]	train-error:0.005743	test-error:0.005744	train-MCC:0.472652	test-MCC:0.416809
[20]	train-error:0.005687	test-error:0.005652	train-MCC:0.473779	test-MCC:0.41783
[21]	train-error:0.00562	test-error:0.005592	train-MCC:0.478876	test-MCC:0.41918
[22]	train-error:0.005599	test-error:0.005542	train-MCC:0.483877	test-MCC:0.420947
[23]	train-error:0.005521	test-error:0.005445	train-MCC:0.484584	test-MCC:0.421386
[24]	train-error:0.005426	test-error:0.005381	train-MCC:0.487854	test-MCC:0.422979
[25]	train-error:0.005334	test-error:0.005271	train-MCC:0.490828	test-MCC:0.425272
[26]	train-error:0.005278	test-error:0.005216	train-MCC:0.49363	test-MCC:0.42579
[27]	train-error:0.005233	test-error:0.005178	train-MCC:0.495648	test-MCC:0.426295
[28]	train-error:0.005206	test-error:0.005157	train-MCC:0.498986	test-MCC:0.430088
[29]	train-error:0.005193	test-error:0.005136	train-MCC:0.500567	test-MCC:0.429714
[30]	train-error:0.005172	test-error:0.005136	train-MCC:0.501945	test-MCC:0.430164
[31]	trai

[38]	train-error:0.005003	test-error:0.005073	train-MCC:0.50771	test-MCC:0.446498
[39]	train-error:0.004981	test-error:0.005064	train-MCC:0.508721	test-MCC:0.446961
[40]	train-error:0.004966	test-error:0.005048	train-MCC:0.511676	test-MCC:0.449032
[41]	train-error:0.004943	test-error:0.005043	train-MCC:0.512665	test-MCC:0.448399
[42]	train-error:0.004912	test-error:0.005031	train-MCC:0.514552	test-MCC:0.450429
[43]	train-error:0.004891	test-error:0.005026	train-MCC:0.515408	test-MCC:0.450236
[44]	train-error:0.004878	test-error:0.005022	train-MCC:0.51714	test-MCC:0.451215
[45]	train-error:0.004866	test-error:0.00501	train-MCC:0.519121	test-MCC:0.449642
[46]	train-error:0.004842	test-error:0.004984	train-MCC:0.521155	test-MCC:0.44964
[47]	train-error:0.00483	test-error:0.004959	train-MCC:0.522032	test-MCC:0.448719
[48]	train-error:0.004819	test-error:0.00495	train-MCC:0.522997	test-MCC:0.448905
[49]	train-error:0.004801	test-error:0.00495	train-MCC:0.52384	test-MCC:0.44885
[50]	train-er

[57]	train-error:0.004704	test-error:0.004895	train-MCC:0.53015	test-MCC:0.449642
[58]	train-error:0.004699	test-error:0.004891	train-MCC:0.531286	test-MCC:0.450428
[59]	train-error:0.00469	test-error:0.00487	train-MCC:0.532084	test-MCC:0.449675
[60]	train-error:0.004686	test-error:0.004862	train-MCC:0.533307	test-MCC:0.449869
[61]	train-error:0.004674	test-error:0.004862	train-MCC:0.534093	test-MCC:0.450428
[62]	train-error:0.004656	test-error:0.004857	train-MCC:0.535033	test-MCC:0.451644
[63]	train-error:0.004649	test-error:0.004845	train-MCC:0.535073	test-MCC:0.450621
[64]	train-error:0.00464	test-error:0.004836	train-MCC:0.535612	test-MCC:0.450283
[65]	train-error:0.004628	test-error:0.004811	train-MCC:0.537843	test-MCC:0.452085
[66]	train-error:0.004619	test-error:0.004807	train-MCC:0.538016	test-MCC:0.450677
[67]	train-error:0.004608	test-error:0.004798	train-MCC:0.539077	test-MCC:0.451682
[68]	train-error:0.004599	test-error:0.004786	train-MCC:0.540219	test-MCC:0.45186
[69]	trai

[76]	train-error:0.004525	test-error:0.004689	train-MCC:0.546611	test-MCC:0.462724
[77]	train-error:0.004522	test-error:0.004667	train-MCC:0.547924	test-MCC:0.463856
[78]	train-error:0.004515	test-error:0.004672	train-MCC:0.549247	test-MCC:0.461007
[79]	train-error:0.00451	test-error:0.004655	train-MCC:0.549758	test-MCC:0.462679
Repeat 0, split 4, test MCC = 0.463, running time = 6.566 min
[0]	train-error:0.005811	test-error:0.005812	train-MCC:0.257943	test-MCC:0.242612
[1]	train-error:0.005811	test-error:0.005812	train-MCC:0.329596	test-MCC:0.312771
[2]	train-error:0.005811	test-error:0.005812	train-MCC:0.357052	test-MCC:0.346443
[3]	train-error:0.005811	test-error:0.005812	train-MCC:0.372482	test-MCC:0.360479
[4]	train-error:0.005811	test-error:0.005812	train-MCC:0.375836	test-MCC:0.36527
[5]	train-error:0.005811	test-error:0.005812	train-MCC:0.387231	test-MCC:0.37039
[6]	train-error:0.005811	test-error:0.005812	train-MCC:0.401312	test-MCC:0.374431
[7]	train-error:0.005811	test-error

[14]	train-error:0.00581	test-error:0.005808	train-MCC:0.453519	test-MCC:0.408602
[15]	train-error:0.00581	test-error:0.005808	train-MCC:0.456712	test-MCC:0.411744
[16]	train-error:0.005808	test-error:0.005808	train-MCC:0.461608	test-MCC:0.415764
[17]	train-error:0.005807	test-error:0.005808	train-MCC:0.462637	test-MCC:0.418164
[18]	train-error:0.005806	test-error:0.005808	train-MCC:0.470239	test-MCC:0.422597
[19]	train-error:0.005804	test-error:0.005808	train-MCC:0.472454	test-MCC:0.428477
[20]	train-error:0.005781	test-error:0.005791	train-MCC:0.476226	test-MCC:0.430435
[21]	train-error:0.00573	test-error:0.005757	train-MCC:0.479585	test-MCC:0.430094
[22]	train-error:0.005674	test-error:0.005711	train-MCC:0.483516	test-MCC:0.431996
[23]	train-error:0.005586	test-error:0.005614	train-MCC:0.486866	test-MCC:0.435059
[24]	train-error:0.005491	test-error:0.005512	train-MCC:0.489352	test-MCC:0.438594
[25]	train-error:0.005385	test-error:0.005381	train-MCC:0.492189	test-MCC:0.440505
[26]	tr

In [ ]:
cv_train_mean = cv_results['train'].mean(axis=1)
cv_train_std = cv_results['train'].std(axis=1)
cv_test_mean = cv_results['test'].mean(axis=1)
cv_test_std = cv_results['test'].std(axis=1)

In [ ]:
plt.figure(figsize=(14, 7))
plt.plot(np.arange(len(cv_train_mean)), cv_train_mean)
plt.fill_between(np.arange(len(cv_train_mean)), cv_train_mean-cv_train_std, cv_train_mean+cv_train_std, alpha=0.5)
plt.plot(np.arange(len(cv_train_mean)), cv_test_mean)
plt.fill_between(np.arange(len(cv_test_mean)), cv_test_mean-cv_test_std, cv_test_mean+cv_test_std, alpha=0.5)
plt.legend(['train', 'test'])

In [ ]:
# Train the model
dtrain = xgb.DMatrix(x_train, label=y_train)
params['seed'] = 28537894
clf = xgb.train(params, dtrain, num_boost_round=60,
               feval=mcc_eval, evals=[(dtrain, 'train')])

y_train_pred = clf.predict(dtrain)

# Find best threshold 
thresholds = np.linspace(0.01, 0.99, 400)
mcc = np.array([matthews_corrcoef(y_train, y_train_pred>thr) for thr in thresholds])
plt.plot(thresholds, mcc)
best_threshold = thresholds[mcc.argmax()]

print('Optimal MCC = {:.3f}'.format(mcc.max()))
print('Optimal threshold = {:.3f}'.format(best_threshold))

### Predict on test data

In [ ]:
dtest = xgb.DMatrix(x_test)
y_test_pred = clf.predict(dtest)
y_test_pred_int = (y_test_pred>best_threshold).astype(int)

sub = pd.read_csv("../../data/sample_submission.csv.zip", index_col=0)
sub["Response"] = y_test_pred_int
sub.to_csv("benchmark_8_submission_cv_5_hash.csv.gz", compression="gzip")

Only With MeanTimeDiff Features:
- LB: `n_estimators=60`: Private MCC = 0.44401, public MCC = 0.43569
- CV: `n_estimators=80`: Score mean = 0.448, std = 0.013
- CV: `n_estimators=60`: Score mean = 0.445491, std = 0.0133, MCC~[0.43218, 0.45880]

Time series features, excluding MeanTimeDiff:
- LB: `n_estimators=60`: Private MCC = 0.46212, public MCC = 0.44761
- CV: `n_estimators=80`: Score mean = 0.454509, std = 0.00586
- CV: `n_estimators=60`: Score mean = 0.4534794, std = 0.0067, MCC~[0.44682, 0.46014]

Time series features, including MeanTimeDiff:
- LB: `n_estimators=60`: Private MCC = 0.43169, public MCC = 0.42776
- CV: `n_estimators=80`: Score mean = 0.452, std = 0.009
- CV: `n_estimators=60`: Score mean = 0.449, std = 0.009, MCC~[0.440109, 0.458029]
- Apparently overfitting since CV is higher than LB

In [ ]:
cv_test_mean[60]+cv_test_std[60]

cv_test_mean[60]-cv_test_std[60]

cv_test_mean[79]

cv_test_std[60]